File: app/Models/Bases/Devise.php
<?php

namespace App\Models\Bases;

use Illuminate\Support\Facades\Http;
use Illuminate\Support\Facades\Auth;
use App\Models\Bases\Ressource;

class Devise extends Ressource
{
    protected $path = 'devise.json';
    // protected $disk = 'storage';
    protected Rate $base;
    public Rate $rate;
    protected $api_url;
    public $data;
    public ?float $value;
    public $rates;
    public $supported;

    public function __construct($value = null, $rate = null)
    {
        parent::__construct();
        $this->value = $value;
        $this->setBase(settings('devise', 'XOF'));
        $this->setCurrent($rate ?? $this->base->code);
        $this->rates = $this->data['rates'] ?? [];
    }

    public function update()
    {
        $this->api_url = 'https://open.er-api.com/v6/latest/' . $this->base->code;
        $response = Http::get($this->api_url);
        if ($response->successful()) {
            $this->data = $response->json();
            if (isset($this->data['rates'])) {
                $this->rates = $this->data['rates'];
                return $this->save();
            }
        }
        return false;
    }

    public function convert(?string $to = null)
    {
        $from = $this->rate->code;
        $to = $to ?? $this->base->code;

        if (!isset($this->rates[$from]) || !isset($this->rates[$to])) {
            return $this;
        }
        $this->value = round(($this->value / $this->rates[$from]) * $this->rates[$to], 2);
        $this->setCurrent($to);
        return $this;
    }

    public function base()
    {
        $this->value = $this->value * ($this->rates[$this->base->code] / $this->rates[$this->getCurrent()->code]);
        $this->setCurrent($this->base->code);
        return $this;
    }

    function format($decimals = 2, $separator = ',', $thousand = ' ')
    {
        return number_format($this->value, $decimals, $separator, $thousand);
    }

    public function getBase()
    {
        return $this->base;
    }

    public function setBase($rate)
    {
        $this->base = $this->makeRate($rate);
    }

    public function getCurrent()
    {
        return $this->rate;
    }

    public function setCurrent($rate)
    {
        $this->rate = $this->makeRate($rate);
    }

    public function makeRate($rate)
    {
        return new Rate($rate);
    }

    public function rateUser()
    {
        return  Auth::user()->parametre['devise'] ?? session('devise', $this->base);
    }

    public function is_valid($rate)
    {
        return array_key_exists($rate, $this->rates);
    }

    public function __toString(): string
    {
        return (string) $this->value;
    }
}


class Rate
{
    public string $name;
    public string $symbol;
    public string $code;

    public static array $names = [
        'EUR' => 'Euro',
        'USD' => 'Dollar',
        'GBP' => 'Livre Sterling',
        'JPY' => 'Yen',
        'INR' => 'Roupie Indienne',
        'XOF' => 'Franc CFA',
        'CHF' => 'Franc Suisse',
        'CRC' => 'Colón Costaricain',
        'PYG' => 'Guarani Paraguyen',
        'UAH' => 'Hryvnia Ukrainienne',
        'RUB' => 'Rouble Russe'
    ];

    public static array $symbols = [
        'EUR' => '€',
        'USD' => '$',
        'GBP' => '£',
        'JPY' => '¥',
        'INR' => '₹',
        'XOF' => 'F',
        'CHF' => '₣',
        'CRC' => '₡',
        'PYG' => '₲',
        'UAH' => '₴',
        'RUB' => '₽',
    ];

    public function __construct($code)
    {
        $this->code = $code;
        $this->name = $this->getName();
        $this->symbol = $this->getSymbol();
    }


    public function getSymbol(): string
    {
        return $this->symbols[$this->code] ?? $this->code;
    }

    public function getName(): string
    {
        return $this->names[$this->code] ?? $this->code;
    }

    public function getCode(): string
    {
        return $this->code;
    }

    public function __toString(): string
    {
        return (string) $this->code;
    }

}

File: app/models/tools/devise.py
from src.core import Storage

class Devise(Storage):
    def __init__(self, value:int=None, rate:str=None):
        self.path = 'tools/devise'
        super().__init__(self.path, format="json", absolute=True, default={})  
        self.value = int(value)
        self.data = self.load() or {}
        self.base = Rate('XOF')
        self.rate = Rate(rate.upper() or self.base.code)
        self.rates = self.data.get('rates', {})
        self.api_url = f"https://open.er-api.com/v6/latest/{self.base.code}"
        self.supported = None

    def update(self):
        import requests 
        response = requests.get(self.api_url)
        if response.status_code == 200:
            self.data = response.json() 
            if 'rates' in self.data:
                self.rates = self.data['rates']
            return self.save()
        return False

    def convert(self, to:str=None):
        from_currency = self.rate.code
        to_currency = to.upper() or self.base.code

        if from_currency not in self.rates or to_currency not in self.rates:
            return self

        self.value = round((self.value / self.rates[from_currency]) * self.rates[to_currency], 2)
        self.set_current(to_currency)
        return self

    def base_currency(self):
        self.value *= (self.rates[self.base.code] / self.rates[self.get_current().code])
        self.set_current(self.base.code)
        return self

    def format_currency(self, decimals=0, separator=',', thousand=' ', devise=False):
        return f"{self.value:,.{decimals}f}".replace(",", "tmp").replace(".", separator).replace("tmp", thousand)

    def get_base(self):
        return self.base

    def set_base(self, rate):
        self.base = Rate(rate)

    def get_current(self):
        return self.rate

    def set_current(self, rate:str):
        self.rate = Rate(rate.upper()) 

    def is_valid(self, rate):
        return rate in self.rates  

    def __str__(self):
        return str(self.value)
    

class Rate:
    names = {
        'EUR': 'Euro',
        'USD': 'Dollar',
        'GBP': 'Livre Sterling',
        'JPY': 'Yen',
        'INR': 'Roupie Indienne',
        'XOF': 'Franc CFA',
        'CHF': 'Franc Suisse',
        'CRC': 'Colón Costaricain',
        'PYG': 'Guarani Paraguyen',
        'UAH': 'Hryvnia Ukrainienne',
        'RUB': 'Rouble Russe'
    }

    symbols = {
        'EUR': '€',
        'USD': '$',
        'GBP': '£',
        'JPY': '¥',
        'INR': '₹',
        'XOF': 'F',
        'CHF': '₣',
        'CRC': '₡',
        'PYG': '₲',
        'UAH': '₴',
        'RUB': '₽',
    }

    def __init__(self, code:str):
        self.code = code.upper()
        self.name = self.get_name()
        self.symbol = self.get_symbol()

    def get_symbol(self):
        return self.symbols.get(self.code, self.code)

    def get_name(self):
        return self.names.get(self.code, self.code)

    def get_code(self):
        return self.code

    def __str__(self):
        return self.code